### ISSUE TITLE
Unable to stream to harbor with ices2

### ISSUE THREAD

**Issue by tatokis** [savonet/liquidsoap#1334]
**Describe the bug**
Ices2 is unable to stream to a harbor instance, failing with `EROR stream/ices_instance_stream Failed initial connect to server:1234 (Login failed: Success)`.
The issue resides in libshout implementing an HTTP OPTIONS query.
From the libshout 2.4.2 changelog: `Support OPTIONS for RFC 2817 TLS mode.`

Specifically, using wireshark we can see liquidsoap responding with an HTTP 404, which causes ices2 to abort.

```
OPTIONS * HTTP/1.1
Host: server:1234
User-Agent: IceS 2.0.2
Connection: Upgrade
Upgrade: TLS/1.0, HTTP/1.1

```
and the response
```
HTTP/1.0 404 Not found
Content-Type: text/html

<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"><head><title>Liquidsoap source harbor</title></head><body><p>This page isn't available.</p></body></html>
```

As a workaround, running a version of ices2 with libshout 2.4.1 allows it to stream to liquidsoap.

**To Reproduce**
Setup a basic harbor input without encryption and try to stream to it with ices2.
```
live = input.harbor("live",port=1234,password="1234",id="live")
```

**Expected behavior**
The HTTP request by libshout shouldn't return a 404, allowing ices2 to stream.

**Version details**
 - OS: Ubuntu 20.04
 - Version: 1.4.2

**Install method**
Opam, including the ssl package.

**Comment by toots** [comment]
Thanks for this report. I'll take care of it shortly.

**Comment by toots** [comment]
Hi,

This is a pretty annoying issue.

Currently, we do not support http connection `Upgrade`. The pattern at the moment is either to have plain harbor via `input.harbor` or SSL harbor via `input.harbor.ssl`. Supporting switching requires some refactorization of the harbor code that I'm not sure that we're ready for.

It is unfortunate that ices switch to requiring SSL connection when it is in fact totally capable of supporting a plain one (so it should be a may require). Does the config allows at least to change this? I haven't found anything in it.

I'm taking this issue off the `1.4.3` release target since I believe its scope is much bigger than what we want to ship for that release..

**Comment by tatokis** [comment]
Apologies for the late reply. I completely missed the notification.

The reason I opened this issue is because ices2 can connect just fine to icecast 2.4.4 without an encrypted connection, nor is my icecast configured to handle encrypted connections.
Using wireshark I can see the whole authentication process, as well as the raw ogg header while using the latest libshout, indicating that it's all plaintext.

This would also explain why there's no option to specify if the connection should be encrypted or not, as it seems to be handled automatically.

If I manually telnet to my server and try to send the OPTIONS request to icecast, it just closes the connection without any reply, which is what I think libshout/ices is expecting if it can't upgrade the connection.

```
$ telnet server 8006
Trying sever_ip...
Connected to server.
Escape character is '^]'.
OPTIONS * HTTP/1.1
Host: server:8006
User-Agent: IceS 2.0.2
Connection: Upgrade
Upgrade: TLS/1.0, HTTP/1.1

Connection closed by foreign host.
```

**Comment by toots** [comment]
Makes a lot of sense, thanks. I was reading from the specs but either I didn't understand it or they implemented it in a less strict way. I'll take care of this shortly.

**Comment by tatokis** [comment]
I was initially confused about it as well, which is why I conducted these tests. After digging a bit more, on [page 5 (section 3.3) of RFC 2817](https://tools.ietf.org/html/rfc2817#section-3.3) it specifies
>    Once the TLS handshake completes successfully, the server MUST
>   continue with the response to the original request. Any TLS handshake
>   failure MUST lead to disconnection, per the TLS error alert
>   specification.

which explains this behaviour.

Thanks for looking into this. Much appreciated!

**Comment by toots** [comment]
Ok, this is fixed. Thank you so much for this report! Couple of take away notes:

* `OPTIONS *` query to upgrade to TLS seems pretty uncommon
* Default response for failed `OPTIONS *` to upgrade to TLS seems undefined: https://tools.ietf.org/html/rfc2817

I chose to use a `405` response with an empty `Allow` header. This makes sense semantically and seems to work with ices/shout. That differs from icecast's response, according to your research, but our harbor is also intended to be a fully features HTTP server so I'd like to keep its implementation as generic as possible. Hopefully this doesn't end-up bitting up back later.

While testing this, I also discovered that libshout does multiple "poking" queries, possibly intended to check for available features. This is really annoying as it creates implementation-specific behavior since responses to these queries are based on icecast's implementation only. Luckily, I was able to make it work with harbor by implementing somewhat reasonable changes, again trying to maintain a generic HTTP server type of implementation.

This should go out with the `1.4.3` release, which will happen in the next day or so.
